android/app/src/main/java/com/google/samples/cronet_sample/MainActivity.java

```diff
@@ -83,49 +87,48 @@ private void onItemsLoadComplete() {
 
     private void setUpToolbar() {
-        Toolbar toolbar =  findViewById(R.id.toolbar);
+        Toolbar toolbar = findViewById(R.id.toolbar);
         setSupportActionBar(toolbar);
         getSupportActionBar().setDisplayShowTitleEnabled(false);
         ((TextView) toolbar.findViewById(R.id.title)).setText(R.string.toolbar_title);
     }
 
     /**
      * This calculates and sets on the UI the latency of loading images with Cronet.
-     * @param cronetLatency
+     *
+     * <p>This method must be thread safe as it can be called from multiple Cronet callbacks
+     * in parallel.
      */
-    public void addCronetLatency(final long cronetLatency) {
-
-        totalLatency += cronetLatency;
-        numberOfImages++;
-
-        if (numberOfImages == ImageRepository.numberOfImages()) {
-            final long averageLatency = totalLatency / numberOfImages;
+    public void onCronetImageLoadSuccessful(long requestLatencyNanos) {
+        CronetMetrics delta = new CronetMetrics(requestLatencyNanos, 1);
+        CronetMetrics newMetrics = metrics.accumulateAndGet(
+                delta,
+                (left, right) -> new CronetMetrics(
+                        left.totalLatencyNanos + right.totalLatencyNanos,
+                        left.numberOfLoadedImages + right.numberOfLoadedImages));
+
+
+        if (newMetrics.numberOfLoadedImages == Math.min(
+                getCronetApplication().imagesToLoadCeiling.get(),
+                ImageRepository.numberOfImages())) {
+            long averageLatencyNanos =
+                    newMetrics.totalLatencyNanos / newMetrics.numberOfLoadedImages;
             android.util.Log.i(TAG,
-                    "All Cronet Requests Complete, the average latency is " + averageLatency);
+                    "All Cronet Requests Complete, the average latency is " + averageLatencyNanos
+                            + " nanos.");
             final TextView cronetTime = findViewById(R.id.cronet_time_label);
-            runOnUiThread(new Runnable() {
-                @Override
-                public void run() {
-                    cronetTime.setText(String.format(getResources()
-                            .getString(R.string.images_loaded), averageLatency));
-                }
-            });
-            this.cronetLatency.set(averageLatency);
+            runOnUiThread(() -> cronetTime.setText(String.format(getResources()
+                    .getString(R.string.images_loaded), averageLatencyNanos)));
         }
     }
 
-    private void setCronetEngine() {
-        // create the Cronet engine, enable caching of HTTP data and
-        // other information like QUIC server information, HTTP/2 protocol and QUIC protocol.
-        cronetEngine = new CronetEngine.Builder(this)
-            .enableHttpCache(CronetEngine.Builder.HTTP_CACHE_IN_MEMORY, 100 * 1024)
-            .enableHttp2(true)
-            .enableQuic(true)
-            .build();
-
+    public CronetApplication getCronetApplication() {
```

Comment:
> Shouldn't this be private since it's not being used by any external files? 

It's used in the view adapter, but can be made package private. Done.

> Also, are you sure we shouldn't cache the cronetEngine/application ie put it in a variable? I'm asking because each time we call getCronetEngine(), it calls getCronetApplication() which in turn calls getApplication();

There's nothing wrong about that, they're just convenience getters to retrieve a shared instance. You could technically store it locally but it's unnecessary, just like you don't store fields accessed by a getter from other objects.